### PR TITLE
[IMP] html_editor: caption button for images

### PR DESCRIPTION
**Pull request by Zynton**
### Purpose

For multiple reasons, users might want to add captions to images:

- **Improved accessibility**: captions add context for all users (cognitive help)
- **Clearer content**: captions help explain the image (more detailed than an alt or title attribute)
- **SEO benefits**: captions help search engines in indexing
- **Clearer design**: captions help keep text and images organized
- **Useful for various users**: such a feature fits different content needs

### Specification

- In the toolbar, when an image is selected, add a "Caption" button next to the "Description" button (in the same button group).
- When the "Caption" button is clicked, if the image has no caption yet, wrap the image in a `figure` element and add a `figcaption` element. The `figcaption` element is editable and focused, with the placeholder text "Write a caption..." (as a hint).
- If the image already has a caption, clicking the "Caption" button focuses the `figcaption` element and puts the caret at the end of its text.
- The `figcaption` element's text is gray.
- When editing the caption, users aren't allowed to add styles, use the toolbar or any Powerbox commands. Merging or splitting the caption is also prohibited. The caption is plain text. It behaves the same way as when editing the filename of a file block (`/file``).
- The feature is not available in the website builder, which will use its own plugin.
- Users can't write in the `figure` element, only in the `figcaption` element.
- When vizualizing the image with the FileViewer (double click on the image), the title shown is the caption when it exists.
- The `<figure><figcaption>...</figcaption></figure>` structure is respected for accessibility purposes.

### Technical details

The caption feature is implemented as an embedded component so it can be completely loose from the editor. This way we ensure the caption remains plain text and can't be merged or split. The user writes in an `input` element in the `figcaption` element, to keep it as plain text. As the caption changes, the `placeholder` attribute of the image is adapted to reflect it. This way, screen readers will still read the caption when selecting the figure (tested with ChromeVox).

-----

In order to make this feature work as expected, this PR reintroduces the concept of editable media, which existed in
`web_editor` but was not moved over to `html_editor`. This is necessary in order for media to be editable when their parent has the `contenteditable` attribute set to `false`. There can be cases where we want to be able to edit an image but not to write in its container. This is the case for instance for the following common structure:

```html
<figure contenteditable="false">
    <img/>
    <figcaption contenteditable="true">Image caption</figcaption>
</figure>
```

We typically only want to write in the caption and not in the figure, but we want to be able to replace the image or transform it.

For such a case, we mark the image with the `o_editable_media` class.

task-4224709

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/html_editor/static/src/core/selection_plugin.js [hunk 1]
```diff
@@ -580,10 +580,11 @@ export class SelectionPlugin extends Plugin {
     /**
      * Returns an array containing all the nodes fully contained in the selection.
      *
+     * @param { EditorSelection } [selection]
      * @returns {Node[]}
      */
-    getSelectedNodes() {
-        const selection = this.getSelectionData().editableSelection;
+    getSelectedNodes(selection) {
+        selection = selection || this.getSelectionData().editableSelection;
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/core/selection_plugin.js:656]: Make it an explicit default parameter

File: addons/html_editor/static/src/core/selection_plugin.js [hunk 2]
```diff
@@ -806,6 +875,10 @@ export class SelectionPlugin extends Plugin {
         return [node, offset];
     }
 
+    isNodeEditable(node) {
+        return node.parentElement?.isContentEditable;
+    }
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/core/selection_plugin.js:880]: The name of this function is confusing IMO. I originally did not realize it was exported as an `editable_nodes_predicates` so I thought this was some local function that the `selection_plugin` was using to determine editability. Considering it's extremely short I would have inlined it in the resources to minimize any confusion.

File: addons/html_editor/static/src/core/selection_plugin.js [hunk 3]
```diff
@@ -457,10 +497,38 @@ export class SelectionPlugin extends Plugin {
...
+            [focusNode, focusOffset] = this.normalizeSelfClosingElement(focusNode, focusOffset);
+            [anchorNode, anchorOffset] = isCollapsed
+                ? [focusNode, focusOffset]
+                : this.normalizeSelfClosingElement(anchorNode, anchorOffset);
+        } else {
+            [focusNode, focusOffset] = this.normalizeCursorPosition(focusNode, focusOffset, "right");
+            [anchorNode, anchorOffset] = isCollapsed
+                ? [focusNode, focusOffset]
+                : this.normalizeCursorPosition(anchorNode, anchorOffset, "left");
+        }
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/core/selection_plugin.js:531]: This is lots of changes in `selection_plugin`. I still think we should not change the actual selection, even when called from `setSelection`.

File: addons/html_editor/static/src/main/media/media_plugin.js [hunk 4]
```diff
@@ -341,4 +344,18 @@ export class MediaPlugin extends Plugin {
...
+        return media.classList.contains(EDITABLE_MEDIA_CLASS);
+    }
+    isNodeEditableMedia(node) {
+        return (isMediaElement(node) || node.nodeName === "IMG") && this.isMediaEditable(node);
+    }
+    isNodeEditable(node) {
+        if (isMediaElement(node) || node.nodeName === "IMG") {
+            return this.isMediaEditable(node);
+        }
+    }
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/main/media/media_plugin.js:364]: Now that I think about it, why didn't we simply put `contenteditable="true"` on the image itself? It's not like we can write inside the image anyway and, conceptually, it could be seen as editing the "content" of the html image, that is the editing the actual image file it points to.

Also, none of these functions are used, but I suppose they will be in later commits.

File: addons/html_editor/static/src/plugin.js [hunk 5]
```diff
@@ -38,7 +38,9 @@ export class Plugin {
     setup() {}
 
     isValidTargetForDomListener(ev) {
-        return !isProtecting(ev.target) && (!isProtected(ev.target) || isUnprotecting(ev.target));
+        return (
+            !isProtecting(ev.target) && (!isProtected(ev.target) || isUnprotecting(ev.target))
+        ) || this.getResource("editable_node_predicates").some(predicate => predicate(ev.target));
```
Comment (dmo-odoo) [inline addons/html_editor/static/src/plugin.js:43]: You're adding a way to bypass all the "protecting"/"protected" checks... :/

File: addons/html_editor/static/src/core/selection_plugin.js (hunk 3 above)
Comment (Zynton) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: Do you want me to add a commit to prevent changing the actual selection when calling `setSelection` or should we create a new task for that?

File: addons/html_editor/static/src/plugin.js (hunk 5 above)
Comment (Zynton) [reply addons/html_editor/static/src/plugin.js:43]: Oops :grimacing:

File: addons/html_editor/static/src/core/selection_plugin.js (hunk 3 above)
Comment (dmo-odoo) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: It depends. Can you make your caption feature work without all the currently proposed changes in `setSelection` and without making the change I'm suggesting?

Comment (Zynton) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: I don't believe the change you're suggesting would affect the feature at all or let me remove my changes in selection_plugin.

Comment (dmo-odoo) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: I don't understand. Isn't all your changes to avoid normalizing because normalizing changes the selection in a way that you don't want? If we don't change the user selection in that case then I don't understand why the other changes would be needed. We need to discuss this on Friday because it's obvious we don't understand each other.

File: addons/html_editor/static/src/main/media/media_plugin.js (hunk 4 above)
Comment (Zynton) [reply addons/html_editor/static/src/main/media/media_plugin.js:364]: For the image to be selected we (currently and in my branch) need the selection to be `<figure>[]<figcaption/></figure>` which is `(figure, 0, figure, 1)` so we'd the figure to be contenteditable, not the image. `(img, 0, img, 0)` wouldn't mean much and would seem collapsed. Our system currently checks whether the node we're referencing in the selection (`figure` in the case of (figure, 0, figure 1)) is contenteditable.
Or did you mean more symbolically, to use the contenteditable attribute instead of a class and keep the rest of my implementation the same?

They will indeed be used in later commits, when the caption feature gets introduced.

File: addons/html_editor/static/src/core/selection_plugin.js (hunk 3 above)
Comment (Zynton) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: `getSelectionData` gets you an `EditorSelection` object, with several versions of the selection in it. One of them is the document selection, which is the one that is actually set in the document. It gets changed when calling `setSelection` if the selection we're trying to set is not editable. I agree that's wrong. Another one is the editable selection, which is for use within editor functions because it's guaranteed to be safe and editable, and it can be different from the document selection. So I don't care about the document selection. What I want is for editor functions to work on the editable selection I need, namely (figure, 0, figure, 1) for my image.
In short, I can make it so that the document selection doesn't change, but I'll still need to change the way the editable selection is corrected.

Comment (dmo-odoo) [reply addons/html_editor/static/src/core/selection_plugin.js:531]: Ok, I see, but that's just wrong. The `editableSelection`, by definition if I understood correctly, _must_ be in editable content. Your proposed changes break that assumption because of your special case.

File: addons/html_editor/static/src/main/media/media_plugin.js (hunk 4 above)
Comment (dmo-odoo) [reply addons/html_editor/static/src/main/media/media_plugin.js:364]: Yes I meant symbolically.

Comment (Zynton) [reply addons/html_editor/static/src/main/media/media_plugin.js:364]: I'm not against it on principle but wouldn't that be even more confusing as it's attached to a separate concept?

Comment (Zynton) [reply addons/html_editor/static/src/main/media/media_plugin.js:364]: Also I'd be weary of side effects, be it from Phoenix or from the browser. You never know ^^'

File: addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        }
+    }
+
+    getCaptionId() {
+        return "" + Math.floor(Math.random() * Date.now());
+    }
+
+    addImageCaption(image, captionText = "", focusInput = true) {
+        image = image || this.dependencies.image.getSelectedImage();
+        if (!image) {
```
Comment (sebgeelen) [inline addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js:105]: Same here, never falsy.

File: addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        const caption = renderToElement("html_editor.EmbeddedCaptionBlueprint", {
+            embeddedProps: JSON.stringify({
+                id: captionId,
+                focusInput,
+            }),
+        });
+        figure.append(caption);
+        // Ensure it's not possible to write inside the figure.
+        figure.setAttribute("contenteditable", "false");
+        // Ensure it's possible to write before and after the figure.
```
Comment (sebgeelen) [inline addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js:147]: should this be aware of the current editor `baseContainer`and use it, instead of always a `p` ?

File: addons/html_editor/static/src/others/embedded_components/backend/caption/caption.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            }
+        });
+        this.observer.observe(this.image, { attributes: true });
+    }
+
+    destroy() {
+        this.observer.disconnect();
+    }
+
+    updateCaption(caption = "", addHistoryStep = true) {
```
Comment (sebgeelen) [inline addons/html_editor/static/src/others/embedded_components/backend/caption/caption.js:52]: related to the this file comment.
I hate this `addHistoryStep = true` parameter. It is such a in hindsight parameter. I would really like to avoid those kind of stuff if possible.

File: addons/html_editor/static/src/plugin.js [hunk 9]
```diff
@@ -42,16 +42,39 @@ export class Plugin {
...
 
+    /**
+     * Add an event listener on the editor's document, and ensure it is removed
+     * when we destroy the editor.
+     *
+     * @param {string} eventName
+     * @param {function(Event):void} fn
+     * @param {boolean} [capture=false] `useCapture` flag of `addEventListener`
+     */
+    addGlobalDomListener(eventName, fn, capture = false) {
```
Comment (sebgeelen) [inline addons/html_editor/static/src/plugin.js:74]: I'm not a big far of the `global` labeling.
In view of the issue we have when the editor is setup in an ifram( website) people could deduce that the `addGlobalDomListener` will listen to even on the main page document, and not the iframe document.

maybe we could name it `addDocumentDomListener` ?

**Review by sebgeelen — COMMENTED** [review]
I mainly focused of the code.
I did not read the tests, so i trust you write correct test cases. :smile:

**Review by Zynton — COMMENTED** [review]
@sebgeelen I added a commit for a significant change made based on your suggestions. If you validate, I'll squash it with the "add image captions" commit. Thanks for your review!

File: addons/html_editor/static/src/plugin.js (hunk 9 above)
Comment (sebgeelen) [reply addons/html_editor/static/src/plugin.js:74]: I see, could you just add a `// @todo use this function to avoid iframe problems` or something to that effect.  
In order to clarify that it is build _also_ as a future potential "fix" for iframe problems.

File: addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+        // Set the caption and its ID.
+        const captionId = this.getCaptionId();
+        image.setAttribute("data-caption-id", captionId);
+        image.setAttribute("data-caption", captionText || "");
+        // Add the caption component.
+        // => <p><figure><img/><figcaption>...</figcaption></figure></p>
+        // or <p><a><figure><img/><figcaption>...</figcaption></figure></a></p>
+        const caption = renderToElement("html_editor.EmbeddedCaptionBlueprint", {
+            embeddedProps: JSON.stringify({
+                id: captionId,
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/html_editor/static/src/others/embedded_components/plugins/caption_plugin/caption_plugin.js:147]: indeed, I overlooked the `JSON.stringify`. :ok_hand:

File: addons/html_editor/static/src/others/embedded_components/backend/caption/caption.js
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/html_editor/static/src/others/embedded_components/backend/caption/caption.js:1]: I like it! 
The sepration of concern clearify the code IMO. ++

**Comment by sebgeelen** [comment]
@robodoo r+

**Comment by Zynton** [comment]
@robodoo rebase-ff
